### ISSUE TITLE
[Bug][Scheduler][Mamba] Track uncomputed align checkpoint visibility after same-step priority preemption

### ISSUE THREAD

**Issue by suntp** [vllm-project/vllm#57580]
This is a scoped reproduction tracker for existing work, not a new root-cause claim.

Related work: #43569 (rollback of eager prefix-cache registrations on uncommitted free) and #56525 (Mamba state-hash publish/commit fencing). The case below is additional CPU regression evidence for those efforts; it does not propose a competing fix or claim that either PR has been validated here.

**Observed behavior**

With `mamba_cache_mode="align"`, prefix caching enabled, 16-token blocks, 14 pool blocks, and no KV connector, priority preemption removes `low` from the final output of a scheduling pass. A later consumer nevertheless obtains its 48-token checkpoint, although only 32 tokens of `low` were covered by previously completed steps in this CPU fixture.

| Policy | Final pressure-step batch | `low` completed position | Consumer prefix hit | State block refcount on acquisition |
|---|---|---:|---:|---|
| `priority` | `high`: 16 tokens; `low` withdrawn | 32 | **48** | 0 → 1; allocation succeeds |
| `fcfs` control | `low`: 16 tokens; `high` preempted | 48 | 48 | 1 → 2; allocation succeeds |

Expected: a request withdrawn before execution must not leave its never-written boundary usable as a computed Mamba checkpoint. Previously completed checkpoints should remain reusable.

The relevant ordering is eager hash registration during `allocate_slots()`, followed by same-pass priority withdrawal, followed by freeing the block without withdrawing that registration. The next-step Mamba guard reset then permits acquisition. A successful reference-count increment does not establish that the checkpoint was computed.

**Environment and evidence limits**

- Executed with Python 3.12.12, `vllm==0.29.0+cpu`, and `torch==2.13.0` on macOS/arm64.
- The final rerun disabled out-of-tree plugins with `VLLM_PLUGINS=''`; after execution, `current_platform` was `vllm.platforms.cpu.CpuPlatform` and no `mlx` modules were loaded.
- This is a CPU scheduler/cache-metadata reproduction. Worker completion is simulated with `ModelRunnerOutput`; no model, GPU, MLX, KV buffers, or connector is used. It does not demonstrate model-output corruption.
- The initial condition is **synthetically seeded metadata for an assumed-valid retained checkpoint at token 64**. Its historical prefill is not executed. Subsequent low/high scheduling, completion updates, preemption, cache lookup, and consumer acquisition use the installed upstream implementation.
- The FCFS control passes for this same fixture; this report does not claim the same trigger under FCFS.
- Latest main `017dced6` was inspected only. This reproduction has **not** been executed against latest main, #43569, or #56525, so this report does not establish whether either proposed fix passes it.

**Self-contained reproduction**

Save the folded script as `priority_checkpoint_repro.py`. It requires the installed vLLM/PyTorch packages above, with no repository test helpers or external files:

```sh
VLLM_PLUGINS='' VLLM_TARGET_DEVICE=cpu VLLM_LOGGING_LEVEL=ERROR .venv/bin/python priority_checkpoint_repro.py
```

<details>
<summary>Single-file CPU reproduction</summary>

```python
"""CPU reproduction for existing vLLM eager-checkpoint/priority-preemption work.

The retained 64-token checkpoint is synthetically seeded and assumed valid.
Worker completion is simulated; no model, GPU, MLX, or connector is used.
"""

import hashlib
import json
import pickle
from importlib.metadata import version
from types import SimpleNamespace

import torch
from vllm import SamplingParams
from vllm.config import (
    CacheConfig,
    ObservabilityConfig,
    ParallelConfig,
    SchedulerConfig,
)
from vllm.v1.core.kv_cache_utils import get_request_block_hasher, init_none_hash
from vllm.v1.core.sched.scheduler import Scheduler
from vllm.v1.kv_cache_interface import (
    FullAttentionSpec,
    KVCacheConfig,
    KVCacheGroupSpec,
    MambaSpec,
)
from vllm.v1.outputs import ModelRunnerOutput
from vllm.v1.request import Request, RequestStatus

BLOCK = 16


def digest(value):
    return hashlib.sha256(pickle.dumps(value)).digest()


def make_scheduler(policy):
    state = MambaSpec(
        block_size=BLOCK,
        shapes=((4, 4),),
        dtypes=(torch.float32,),
        mamba_cache_mode="align",
        num_prefill_checkpoint_blocks=0,
    )
    kv = KVCacheConfig(
        num_blocks=14,
        kv_cache_tensors=[],
        kv_cache_groups=[
            KVCacheGroupSpec(
                ["attention"],
                FullAttentionSpec(
                    block_size=BLOCK, num_kv_heads=1, head_size=2, dtype=torch.float32
                ),
            ),
            KVCacheGroupSpec(["state"], state),
        ],
        prefix_cache_retention_interval=0,
    )
    cache = CacheConfig(
        block_size=BLOCK,
        mamba_block_size=BLOCK,
        mamba_cache_mode="align",
        enable_prefix_caching=True,
    )
    cache.num_gpu_blocks = 14
    config = SimpleNamespace(
        scheduler_config=SchedulerConfig(
            max_model_len=512,
            is_encoder_decoder=False,
            max_num_batched_tokens=64,
            max_num_seqs=4,
            async_scheduling=False,
            enable_chunked_prefill=True,
            long_prefill_token_threshold=BLOCK,
            policy=policy,
        ),
        cache_config=cache,
        parallel_config=ParallelConfig(),
        observability_config=ObservabilityConfig(),
        model_config=SimpleNamespace(
            uses_mrope=False,
            uses_xdrope=False,
            is_encoder_decoder=False,
            is_diffusion=False,
            max_model_len=512,
            enable_return_routed_experts=False,
            return_sampling_mask=False,
        ),
        lora_config=None,
        kv_events_config=None,
        is_mm_encoder_only=False,
        kv_transfer_config=None,
        ec_transfer_config=None,
        ec_manager_config=SimpleNamespace(get_encoder_cache_manager_obj=lambda: None),
        speculative_config=None,
        num_speculative_tokens=0,
        num_lookahead_tokens=0,
        max_in_flight_tokens=64,
        max_concurrent_batches=1,
        use_v2_model_runner=False,
    )
    return Scheduler(
        config,
        kv,
        structured_output_manager=SimpleNamespace(should_advance=lambda *a, **k: False),
        block_size=BLOCK,
        hash_block_size=BLOCK,
        mm_registry=SimpleNamespace(supports_multimodal_inputs=lambda _: False),
    )


def request(name, tokens, offset=0, priority=0):
    return Request(
        request_id=name,
        prompt_token_ids=list(range(offset, offset + tokens)),
        sampling_params=SamplingParams(temperature=0, max_tokens=128, ignore_eos=True),
        pooling_params=None,
        block_hasher=get_request_block_hasher(BLOCK, digest),
        priority=priority,
    )


def complete(scheduler, step):
    # Simulate completion only for requests in the final SchedulerOutput.
    ids = list(step.num_scheduled_tokens)
    scheduler.update_from_output(
        step,
        ModelRunnerOutput(
            req_ids=ids,
            req_id_to_index={rid: i for i, rid in enumerate(ids)},
            sampled_token_ids=[
                [] if scheduler.requests[rid].is_prefill_chunk else [1001]
                for rid in ids
            ],
        ),
    )


def run(policy):
    scheduler = make_scheduler(policy)
    manager = scheduler.kv_cache_manager
    owner = request("retained_owner", 65)
    manager.new_step_starts()
    assert manager.allocate_slots(owner, 64) is not None
    # This is the synthetic initial condition, not a historical model forward.
    owner.num_computed_tokens = 64
    owner.status = RequestStatus.WAITING_FOR_STREAMING_REQ
    owner.resumable = True

    low = request("low", 49, priority=10)
    high = request("high", 33, offset=1000, priority=0)
    scheduler.add_request(low)
    first = scheduler.schedule()
    assert first.num_scheduled_tokens == {"low": 16}
    complete(scheduler, first)
    scheduler.add_request(high)
    second = scheduler.schedule()
    assert second.num_scheduled_tokens == {"low": 16, "high": 16}
    complete(scheduler, second)

    completed_low = low.num_computed_tokens
    pressure = scheduler.schedule()
    complete(scheduler, pressure)
    completed_low += pressure.num_scheduled_tokens.get("low", 0)
    manager.new_step_starts()
    consumer = request("consumer", 49)
    hits, cached_tokens, *_ = manager.get_computed_blocks(consumer)
    state = hits.blocks[1][-1]
    refs_before = state.ref_cnt
    allocated = manager.allocate_slots(
        consumer, 1, num_new_computed_tokens=cached_tokens, new_computed_blocks=hits
    )
    result = {
        "policy": policy,
        "pressure_scheduled": pressure.num_scheduled_tokens,
        "preempted": sorted(pressure.preempted_req_ids),
        "completed_low_tokens": completed_low,
        "consumer_cached_tokens": cached_tokens,
        "state_refs": [refs_before, state.ref_cnt],
        "consumer_allocated": allocated is not None,
    }
    print(json.dumps(result))
    assert allocated is not None
    if policy == "priority":
        assert pressure.preempted_req_ids == {"low"}
        assert (completed_low, cached_tokens, refs_before, state.ref_cnt) == (
            32,
            48,
            0,
            1,
        )
    else:
        assert (completed_low, cached_tokens) == (48, 48)


if __name__ == "__main__":
    print(json.dumps({"vllm": version("vllm"), "torch": version("torch")}))
    init_none_hash(digest)
    for policy in ("priority", "fcfs"):
        run(policy)
```

</details>

<details>
<summary>Observed stdout (exit code 0)</summary>

```text
{"vllm": "0.29.0+cpu", "torch": "2.13.0"}
{"policy": "priority", "pressure_scheduled": {"high": 16}, "preempted": ["low"], "completed_low_tokens": 32, "consumer_cached_tokens": 48, "state_refs": [0, 1], "consumer_allocated": true}
{"policy": "fcfs", "pressure_scheduled": {"low": 16}, "preempted": ["high"], "completed_low_tokens": 48, "consumer_cached_tokens": 48, "state_refs": [1, 2], "consumer_allocated": true}
```

</details>

The requested outcome is to track this canceled-producer Mamba case alongside #43569/#56525 and retain it as regression coverage, while preserving valid completed-checkpoint reuse and the passing control. Fix validation on a current upstream checkout and a fixture driven from a cold scheduler without the synthetic initial checkpoint remain follow-up work.

Prepared with AI assistance. No production patch or model-level validation is included in this report.

<details>
<summary>Environment collection</summary>

Executed vLLM version: `0.29.0+cpu` (also printed by the reproducer). Current-main `vllm/collect_env.py` fails on this macOS host at the Linux-only `get_pkg_version()` assertion; the successful `torch.utils.collect_env` output follows.

```text
Collecting environment information...
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.6.2 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.0.123.102)
CMake version: version 4.4.3
Libc version: N/A

Python version: 3.12.12 (main, Oct  9 2025, 11:07:00) [Clang 17.0.0 (clang-1700.6.3.2)] (64-bit runtime)
Python platform: macOS-26.6.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M5 Pro

Versions of relevant libraries:
[pip3] mypy==2.3.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.5
[pip3] torch==2.13.0
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.16.0
[pip3] torchvision==0.28.0
[conda] No relevant packages

<frozen runpy>:128: RuntimeWarning: 'torch.utils.collect_env' found in sys.modules after import of package 'torch.utils', but prior to execution of 'torch.utils.collect_env'; this may result in unpredictable behaviour
```

</details>

**Comment by debanshd** [comment]
Hi, I am investigating this issue and working on a fix with unit tests. I will submit a pull request once verification is complete.
